app/services/set_user_preferences.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+# This class handles both the creating and updating of a user's UserPreferences.
+# It takes care of updating because, for each request, we are deleting all
+# of the user's associated UserPreference records, and creating new ones.
+
+# Also note that it accepts and sets multiple preference/choices combinations in
+# each request.  This is so that the end user can make a number of preference decisions
+# in one page, and the FE can send over just one request, in order to create all of the
+# UserPreference records for all of the users decisions.
+#
+# This class is responsible for:
+#   - accepting an Account record and an array of preference/choice pairings
+#   - destroying any associated existing UserPreference records
+#   - creating new UserPreference records based on the request
+#   - returning the successfully created pairings that were requested
+#
```

Comment:
I’ve taken another look at refactoring this and honestly, I think that keeping it in one class is the way to go.

The challenge is the requirements.  There are so many requirements for what has to happen in this endpoint, one of the most challenging ones being a loop inside a loop.  We have to accept an array of request data, and each piece inside of it contains another array.  And within each of these loops, we need to take various actions.  

Because of that, any refactoring is going to require a lot more code, and likely duplication.  Normally I would agree that we should refactor it, but in this case, it feels like pulling things out of this class are going to cause more indirection and confusion, then not.